{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bd7e5c10-f631-4ea3-920b-6ac71e677e79",
   "metadata": {},
   "source": [
    "## Case-study transformation\n",
    "\n",
    "### Import data and prepare source schemas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0008bcb1-5e82-469c-b71b-2081bf1cb018",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.11.4 (main, Nov  7 2023, 16:51:51) [GCC 11.4.0]\n",
      "2025-01-07 09:27:36\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 10:27:46,345\tINFO util.py:154 -- Missing packages: ['ipywidgets']. Run `pip install -U ipywidgets`, then restart the notebook server for rich notebook output.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 850 ms, sys: 2.36 s, total: 3.21 s\n",
      "Wall time: 10.4 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "###################################################################################################\n",
    "### IMPORTS AND CONSTANTS\n",
    "###################################################################################################\n",
    "from time import gmtime, strftime\n",
    "from datetime import date, datetime\n",
    "import sys\n",
    "print(sys.version)\n",
    "print(strftime(\"%Y-%m-%d %H:%M:%S\", gmtime()))\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import whyqd as qd\n",
    "\n",
    "SCHEMA_DESTINATION_PATH = \"whyqd_schemas/fairtracks-convenience-schema.SCHEMA\"\n",
    "DATASOURCE_PATH = \"whyqd_schemas/Meta-Data 49 Diatom Protein Coding Gene Annotations.xlsx\"\n",
    "MIMETYPE = \"xlsx\"\n",
    "SCHEMA_SOURCE = {}\n",
    "DATA_SOURCE = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6241a6be-ef96-4274-80d7-88c37d404c02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-07 09:54:36\n",
      "Failed to delete /tmp/systemd-private-305f46beac1e464c9a3b33c389efa586-systemd-logind.service-YD8DhN. Reason: [Errno 13] Permission denied: '/tmp/systemd-private-305f46beac1e464c9a3b33c389efa586-systemd-logind.service-YD8DhN'\n",
      "Failed to delete /tmp/.X11-unix. Reason: [Errno 30] Read-only file system: 'X0'\n",
      "Failed to delete /tmp/.XIM-unix. Reason: [Errno 1] Operation not permitted: '/tmp/.XIM-unix'\n",
      "Failed to delete /tmp/.Test-unix. Reason: [Errno 1] Operation not permitted: '/tmp/.Test-unix'\n",
      "Failed to delete /tmp/systemd-private-305f46beac1e464c9a3b33c389efa586-systemd-resolved.service-XKhcBt. Reason: [Errno 13] Permission denied: '/tmp/systemd-private-305f46beac1e464c9a3b33c389efa586-systemd-resolved.service-XKhcBt'\n",
      "Failed to delete /tmp/systemd-private-305f46beac1e464c9a3b33c389efa586-systemd-timesyncd.service-yv0SRe. Reason: [Errno 13] Permission denied: '/tmp/systemd-private-305f46beac1e464c9a3b33c389efa586-systemd-timesyncd.service-yv0SRe'\n",
      "Failed to delete /tmp/.ICE-unix. Reason: [Errno 1] Operation not permitted: '/tmp/.ICE-unix'\n",
      "Failed to delete /tmp/snap-private-tmp. Reason: [Errno 13] Permission denied: '/tmp/snap-private-tmp'\n",
      "Failed to delete /tmp/.font-unix. Reason: [Errno 1] Operation not permitted: '/tmp/.font-unix'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 10:54:44,226\tINFO worker.py:1821 -- Started a local Ray instance.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 960 ms, sys: 609 ms, total: 1.57 s\n",
      "Wall time: 26.6 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'input_data_accession_numbers': Schema: `input_data_accession_numbers`,\n",
       " 'entap_functional_annotation_sum': Schema: `entap_functional_annotation_sum`,\n",
       " 'software_and_container_versions': Schema: `software_and_container_versions`,\n",
       " 'annotations_descriptive_statist': Schema: `annotations_descriptive_statist`,\n",
       " 'repeats': Schema: `repeats`,\n",
       " 'busco_scores': Schema: `busco_scores`,\n",
       " 'omark_scores': Schema: `omark_scores`,\n",
       " 'busco_version_dependencies': Schema: `busco_version_dependencies`}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "###################################################################################################\n",
    "### IMPORT SOURCES\n",
    "###################################################################################################\n",
    "print(strftime(\"%Y-%m-%d %H:%M:%S\", gmtime()))\n",
    "\n",
    "# Get the destination schema\n",
    "SCHEMA_DESTINATION = qd.SchemaDefinition(source=SCHEMA_DESTINATION_PATH)\n",
    "# Get data sources from multiple sheets and derive the source schemas for each\n",
    "datasource = qd.DataSourceDefinition()\n",
    "datasource.derive_model(source=DATASOURCE_PATH, mimetype=MIMETYPE)\n",
    "for ds in datasource.get:\n",
    "    # We separate out the individual sheets to transform independently\n",
    "    DATA_SOURCE[ds.sheet_name.lower()] = ds\n",
    "    schema_source = qd.SchemaDefinition(source={\n",
    "          \"name\": ds.sheet_name.lower()\n",
    "    })\n",
    "    schema_source.derive_model(data=ds)\n",
    "    SCHEMA_SOURCE[ds.sheet_name.lower()] = schema_source\n",
    "\n",
    "# Show source schemas\n",
    "SCHEMA_SOURCE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96e416ea-6efb-496f-8efe-8840b68948c2",
   "metadata": {},
   "source": [
    "### Prepare crosswalks for each source schema\n",
    "\n",
    "The joining field across the useable sheets are some variation of `Species name`. However, while useable data are scattered throughout the sheets, only a subset are addressable via a `whyqd` crosswalk while the rest would need to be added in as a `NEW` action.\n",
    "\n",
    "Useable schemas:\n",
    "- input_data_accession_numbers\n",
    "- entap_functional_annotation_sum\n",
    "- annotations_descriptive_statist\n",
    "- repeats\n",
    "- busco_scores\n",
    "- omark_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8a454984-01cd-45e3-881c-4e4b3a9b274e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-07 09:41:52\n",
      "-----whyqd_schemas/sheet1.CROSSWALK-----\n",
      "SELECT > 'local_id' < ['Species name']\n",
      "SELECT > 'track_assembly_id' < ['Genome assembly accession number']\n",
      "NEW > 'track_feature' < ['http://purl.obolibrary.org/obo/SO_0000704']\n",
      "COLLATE > 'experiment_evidence_global_id' < ['RNAseq Libraries used for annotation','Legacy genome assembly accession number or protein source (protein sequence source used for annotation)','OrthoDB Proteins']\n",
      "COLLATE > 'experiment_annotation_method_name' < ['Annotation method']\n",
      "COLLATE > 'study_publication' < ['Data Zenodo DOI','Publication with Methods Descripton']\n",
      "SELECT > 'track_file_url' < ['Data File on Zenodo']\n",
      "SELECT > 'track_file_name' < ['Data File on Zenodo']\n",
      "NEW > 'study_study_name' < ['Annotation of protein-coding genes in 49 diatom genomes from the Bacillariophyta clade']\n",
      "NEW > 'study_contact_name' < ['Katharina Hoff']\n",
      "NEW > 'study_contact_email' < ['katharina.hoff@uni-greifswald.de']\n",
      "NEW > 'study_contact_orcid' < ['orcid:0000-0002-7333-8390']\n",
      "NEW > 'experiment_technique' < ['http://edamontology.org/operation_0362']\n",
      "NEW > 'track_file_format' < ['http://edamontology.org/format_1975']\n",
      "-----whyqd_schemas/sheet2.CROSSWALK-----\n",
      "SELECT > 'local_id' < ['Species name']\n",
      "COLLATE > 'track_evaluation_method_value' < ['BUSCO scores genome level','BUSCO scores protein level']\n",
      "-----whyqd_schemas/sheet3.CROSSWALK-----\n",
      "SELECT > 'local_id' < ['Species']\n",
      "NEW > 'experiment_compute_protocol_description' < ['https://docs.google.com/spreadsheets/d/1_1wKibSXbi8MQCTqpK03ytuQH71yn28xKuRS1i-mVSw/edit?gid=52185137#gid=52185137']\n",
      "-----whyqd_schemas/sheet4.CROSSWALK-----\n",
      "SELECT > 'local_id' < ['Species name']\n",
      "COLLATE > 'track_evaluation_method_value' < ['OMArk sum']\n",
      "CPU times: user 22.7 ms, sys: 0 ns, total: 22.7 ms\n",
      "Wall time: 29.6 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "###################################################################################################\n",
    "### GET THE EXISTING CROSSWALKS\n",
    "###################################################################################################\n",
    "print(strftime(\"%Y-%m-%d %H:%M:%S\", gmtime()))\n",
    "\n",
    "CROSSWALK_SOURCES = [f\"whyqd_schemas/sheet{n+1}.CROSSWALK\" for n in range(4)]\n",
    "for CROSSWALK_SOURCE in CROSSWALK_SOURCES:\n",
    "    print(f\"-----{CROSSWALK_SOURCE}-----\")\n",
    "    crosswalk = qd.CrosswalkDefinition(crosswalk=CROSSWALK_SOURCE)\n",
    "    for a in crosswalk.actions.get_all():\n",
    "        print(a.script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "ea307f48-1930-42b5-938f-9e1f5056e091",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-01-07 11:13:01\n",
      "CPU times: user 863 ms, sys: 290 ms, total: 1.15 s\n",
      "Wall time: 4.39 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "###################################################################################################\n",
    "### PREPARE AND RUN CROSSWALKS\n",
    "###################################################################################################\n",
    "print(strftime(\"%Y-%m-%d %H:%M:%S\", gmtime()))\n",
    "\n",
    "TRANSFORM_DATA = []\n",
    "# Describe the crosswalks\n",
    "CROSSWALK_SCRIPTS = {\n",
    "    \"input_data_accession_numbers\": [\n",
    "        \"SELECT > 'local_id' < ['Species name']\",\n",
    "        \"SELECT > 'track_assembly_id' < ['Genome assembly accession number']\",\n",
    "        \"NEW > 'track_feature' < ['http://purl.obolibrary.org/obo/SO_0000704']\",\n",
    "        \"COLLATE > 'experiment_evidence_global_id' < ['RNAseq Libraries used for annotation','Legacy genome assembly accession number or protein source (protein sequence source used for annotation)','OrthoDB Proteins']\",\n",
    "        \"COLLATE > 'experiment_annotation_method_name' < ['Annotation method']\",\n",
    "        \"COLLATE > 'study_publication' < ['Data Zenodo DOI','Publication with Methods Descripton']\",\n",
    "        \"SELECT > 'track_file_url' < ['Data File on Zenodo']\",\n",
    "        \"SELECT > 'track_file_name' < ['Data File on Zenodo']\",\n",
    "        \"NEW > 'study_study_name' < ['Annotation of protein-coding genes in 49 diatom genomes from the Bacillariophyta clade']\",\n",
    "        \"NEW > 'study_contact_name' < ['Katharina Hoff']\",\n",
    "        \"NEW > 'study_contact_email' < ['katharina.hoff@uni-greifswald.de']\",\n",
    "        \"NEW > 'study_contact_orcid' < ['orcid:0000-0002-7333-8390']\",\n",
    "        \"NEW > 'experiment_technique' < ['http://edamontology.org/operation_0362']\",\n",
    "        \"NEW > 'track_file_format' < ['http://edamontology.org/format_1975']\",\n",
    "    ],\n",
    "    \"entap_functional_annotation_sum\": [\n",
    "        \"SELECT > 'local_id' < ['Species']\",\n",
    "        \"NEW > 'experiment_compute_protocol_description' < ['https://docs.google.com/spreadsheets/d/1_1wKibSXbi8MQCTqpK03ytuQH71yn28xKuRS1i-mVSw/edit?gid=52185137#gid=52185137']\",\n",
    "    ],\n",
    "    \"busco_scores\": [\n",
    "        \"SELECT > 'local_id' < ['Species name']\",\n",
    "        \"COLLATE > 'track_evaluation_method_value' < ['BUSCO scores genome level','BUSCO scores protein level']\",\n",
    "    ],\n",
    "    \"omark_scores\": [\n",
    "        \"SELECT > 'local_id' < ['Species name']\",\n",
    "        \"COLLATE > 'track_evaluation_method_value' < ['OMArk sum']\",\n",
    "    ],\n",
    "}\n",
    "\n",
    "for key, cw_script in CROSSWALK_SCRIPTS.items():\n",
    "    # Define a Crosswalk\n",
    "    crosswalk = qd.CrosswalkDefinition()\n",
    "    crosswalk.set(schema_source=SCHEMA_SOURCE[key], schema_destination=SCHEMA_DESTINATION)\n",
    "    crosswalk.actions.add_multi(terms=cw_script)\n",
    "    # Transform a data source\n",
    "    transform = qd.TransformDefinition(crosswalk=crosswalk, data_source=DATA_SOURCE[key])\n",
    "    transform.process()\n",
    "    TRANSFORM_DATA.append(transform.data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "529d7cdd-d51b-45d9-83eb-2e94bd90e6e4",
   "metadata": {},
   "source": [
    "### Merge the data into a single table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f97cf62-2177-4603-ac50-9e55ccb904dc",
   "metadata": {},
   "source": [
    "`track_evaluation_method_value` is required to be an array (hence the `COLLATE` action). However, to merge these, we have two columns with the same name. We need to treat this as a special case.\n",
    "\n",
    "- Restructure the `omark_scores` field, since this has only a single value, but is structured as `[[x1], [x2], [x3], ...]`.\n",
    "- Concatenate it with the `busco_scores` column and update the original column.\n",
    "- Assign a new `track_evaluation_method_name` with the static value `['BUSCO scores genome level','BUSCO scores protein level', 'OMArk sum']`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "54432e7a-65fe-4be8-8117-b5e1092f3018",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import modin.pandas as pd\n",
    "from itertools import chain\n",
    "\n",
    "# 1. Merge `input_data_accession_numbers` and `entap_functional_annotation_sum`\n",
    "df = pd.merge(TRANSFORM_DATA[0], TRANSFORM_DATA[1], how=\"outer\", on=\"local_id\")\n",
    "\n",
    "# 2. Process `busco_scores` and `omark_scores`, then merge with the merged dataframe\n",
    "# First get the column values as arrays\n",
    "busco_values = TRANSFORM_DATA[2][\"track_evaluation_method_value\"].to_numpy()\n",
    "omark_values = TRANSFORM_DATA[3][\"track_evaluation_method_value\"].to_numpy()\n",
    "# Unpack and get the T value of the `omark` values\n",
    "omark_values = np.array([list(chain(*omark_values.tolist()))])\n",
    "# Concatenate into a new list and assign\n",
    "TRANSFORM_DATA[2][\"track_evaluation_method_value\"] = np.concatenate((busco_values.tolist(), omark_values.T), axis=1).tolist()\n",
    "df = pd.merge(df, TRANSFORM_DATA[2], how=\"outer\", on=\"local_id\")\n",
    "\n",
    "# 3. Add in missing `name` array fields, as required\n",
    "# `experiment_evidence_global_id` requires a corresponding `experiment_evidence_evidence_type` with value `['RNA', 'proteins', 'protein database']`\n",
    "df[\"experiment_evidence_evidence_type\"] = np.tile([\"RNA\", \"proteins\", \"protein database\"], (len(df), 1)).tolist()\n",
    "# `track_evaluation_method_value` requires a corresponding `track_evaluation_method_name` with value `['BUSCO scores genome level','BUSCO scores protein level', 'OMArk sum']`\n",
    "df[\"track_evaluation_method_name\"] = np.tile([\"BUSCO scores genome level\",\"BUSCO scores protein level\", \"OMArk sum\"], (len(df), 1)).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "847dc489-c03f-4630-a394-d158a8b5140b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>local_id</th>\n",
       "      <th>study_study_name</th>\n",
       "      <th>study_publication</th>\n",
       "      <th>study_contact_name</th>\n",
       "      <th>study_contact_email</th>\n",
       "      <th>study_contact_orcid</th>\n",
       "      <th>experiment_technique</th>\n",
       "      <th>track_assembly_id</th>\n",
       "      <th>track_file_name</th>\n",
       "      <th>track_file_url</th>\n",
       "      <th>track_file_format</th>\n",
       "      <th>track_feature</th>\n",
       "      <th>experiment_annotation_method_name</th>\n",
       "      <th>experiment_evidence_global_id</th>\n",
       "      <th>experiment_compute_protocol_description</th>\n",
       "      <th>track_evaluation_method_value</th>\n",
       "      <th>experiment_evidence_evidence_type</th>\n",
       "      <th>track_evaluation_method_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Asterionella formosa</td>\n",
       "      <td>Annotation of protein-coding genes in 49 diato...</td>\n",
       "      <td>[10.5281/zenodo.14040071, https://doi.org/10.4...</td>\n",
       "      <td>Katharina Hoff</td>\n",
       "      <td>katharina.hoff@uni-greifswald.de</td>\n",
       "      <td>orcid:0000-0002-7333-8390</td>\n",
       "      <td>http://edamontology.org/operation_0362</td>\n",
       "      <td>GCA_002256025.1</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Aste...</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Aste...</td>\n",
       "      <td>http://edamontology.org/format_1975</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SO_0000704</td>\n",
       "      <td>[BRAKER3]</td>\n",
       "      <td>[SRR5749612, None, Viridiplantae+Stramenopiles...</td>\n",
       "      <td>https://docs.google.com/spreadsheets/d/1_1wKib...</td>\n",
       "      <td>[C:95.0%[S:94.0%,D:1.0%],F:1.0%,M:4.0%,n:100,E...</td>\n",
       "      <td>[RNA, proteins, protein database]</td>\n",
       "      <td>[BUSCO scores genome level, BUSCO scores prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Asterionellopsis glacialis</td>\n",
       "      <td>Annotation of protein-coding genes in 49 diato...</td>\n",
       "      <td>[10.5281/zenodo.14040071, https://doi.org/10.4...</td>\n",
       "      <td>Katharina Hoff</td>\n",
       "      <td>katharina.hoff@uni-greifswald.de</td>\n",
       "      <td>orcid:0000-0002-7333-8390</td>\n",
       "      <td>http://edamontology.org/operation_0362</td>\n",
       "      <td>GCA_014885115.2</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Aste...</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Aste...</td>\n",
       "      <td>http://edamontology.org/format_1975</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SO_0000704</td>\n",
       "      <td>[BRAKER2]</td>\n",
       "      <td>[None, None, Viridiplantae+Stramenopiles (http...</td>\n",
       "      <td>https://docs.google.com/spreadsheets/d/1_1wKib...</td>\n",
       "      <td>[C:92.0%[S:92.0%,D:0.0%],F:1.0%,M:7.0%,n:100,E...</td>\n",
       "      <td>[RNA, proteins, protein database]</td>\n",
       "      <td>[BUSCO scores genome level, BUSCO scores prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bacterosira constricta</td>\n",
       "      <td>Annotation of protein-coding genes in 49 diato...</td>\n",
       "      <td>[10.5281/zenodo.14040071, https://doi.org/10.4...</td>\n",
       "      <td>Katharina Hoff</td>\n",
       "      <td>katharina.hoff@uni-greifswald.de</td>\n",
       "      <td>orcid:0000-0002-7333-8390</td>\n",
       "      <td>http://edamontology.org/operation_0362</td>\n",
       "      <td>GCA_037356235.1</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Bact...</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Bact...</td>\n",
       "      <td>http://edamontology.org/format_1975</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SO_0000704</td>\n",
       "      <td>[BRAKER3]</td>\n",
       "      <td>[SRR18733610, None, Viridiplantae+Stramenopile...</td>\n",
       "      <td>https://docs.google.com/spreadsheets/d/1_1wKib...</td>\n",
       "      <td>[C:95.0%[S:95.0%,D:0.0%],F:2.0%,M:3.0%,n:100,E...</td>\n",
       "      <td>[RNA, proteins, protein database]</td>\n",
       "      <td>[BUSCO scores genome level, BUSCO scores prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chaetoceros muellerii</td>\n",
       "      <td>Annotation of protein-coding genes in 49 diato...</td>\n",
       "      <td>[10.5281/zenodo.14040071, https://doi.org/10.4...</td>\n",
       "      <td>Katharina Hoff</td>\n",
       "      <td>katharina.hoff@uni-greifswald.de</td>\n",
       "      <td>orcid:0000-0002-7333-8390</td>\n",
       "      <td>http://edamontology.org/operation_0362</td>\n",
       "      <td>GCA_019693545.1</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Chae...</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Chae...</td>\n",
       "      <td>http://edamontology.org/format_1975</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SO_0000704</td>\n",
       "      <td>[BRAKER3]</td>\n",
       "      <td>[SRR8647946,SRR8647948,SRR8647949,SRR8647950,S...</td>\n",
       "      <td>https://docs.google.com/spreadsheets/d/1_1wKib...</td>\n",
       "      <td>[C:97.0%[S:96.0%,D:1.0%],F:1.0%,M:2.0%,n:100, ...</td>\n",
       "      <td>[RNA, proteins, protein database]</td>\n",
       "      <td>[BUSCO scores genome level, BUSCO scores prote...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Conticribra guillardii</td>\n",
       "      <td>Annotation of protein-coding genes in 49 diato...</td>\n",
       "      <td>[10.5281/zenodo.14040071, https://doi.org/10.4...</td>\n",
       "      <td>Katharina Hoff</td>\n",
       "      <td>katharina.hoff@uni-greifswald.de</td>\n",
       "      <td>orcid:0000-0002-7333-8390</td>\n",
       "      <td>http://edamontology.org/operation_0362</td>\n",
       "      <td>GCA_036939335.1</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Cont...</td>\n",
       "      <td>https://zenodo.org/records/14040072/files/Cont...</td>\n",
       "      <td>http://edamontology.org/format_1975</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SO_0000704</td>\n",
       "      <td>[BRAKER3]</td>\n",
       "      <td>[SRR18733581, None, Viridiplantae+Stramenopile...</td>\n",
       "      <td>https://docs.google.com/spreadsheets/d/1_1wKib...</td>\n",
       "      <td>[C:99.0%[S:98.0%,D:1.0%],F:0.0%,M:1.0%,n:100,E...</td>\n",
       "      <td>[RNA, proteins, protein database]</td>\n",
       "      <td>[BUSCO scores genome level, BUSCO scores prote...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     local_id  \\\n",
       "0        Asterionella formosa   \n",
       "1  Asterionellopsis glacialis   \n",
       "2      Bacterosira constricta   \n",
       "3       Chaetoceros muellerii   \n",
       "4      Conticribra guillardii   \n",
       "\n",
       "                                    study_study_name  \\\n",
       "0  Annotation of protein-coding genes in 49 diato...   \n",
       "1  Annotation of protein-coding genes in 49 diato...   \n",
       "2  Annotation of protein-coding genes in 49 diato...   \n",
       "3  Annotation of protein-coding genes in 49 diato...   \n",
       "4  Annotation of protein-coding genes in 49 diato...   \n",
       "\n",
       "                                   study_publication study_contact_name  \\\n",
       "0  [10.5281/zenodo.14040071, https://doi.org/10.4...     Katharina Hoff   \n",
       "1  [10.5281/zenodo.14040071, https://doi.org/10.4...     Katharina Hoff   \n",
       "2  [10.5281/zenodo.14040071, https://doi.org/10.4...     Katharina Hoff   \n",
       "3  [10.5281/zenodo.14040071, https://doi.org/10.4...     Katharina Hoff   \n",
       "4  [10.5281/zenodo.14040071, https://doi.org/10.4...     Katharina Hoff   \n",
       "\n",
       "                study_contact_email        study_contact_orcid  \\\n",
       "0  katharina.hoff@uni-greifswald.de  orcid:0000-0002-7333-8390   \n",
       "1  katharina.hoff@uni-greifswald.de  orcid:0000-0002-7333-8390   \n",
       "2  katharina.hoff@uni-greifswald.de  orcid:0000-0002-7333-8390   \n",
       "3  katharina.hoff@uni-greifswald.de  orcid:0000-0002-7333-8390   \n",
       "4  katharina.hoff@uni-greifswald.de  orcid:0000-0002-7333-8390   \n",
       "\n",
       "                     experiment_technique track_assembly_id  \\\n",
       "0  http://edamontology.org/operation_0362   GCA_002256025.1   \n",
       "1  http://edamontology.org/operation_0362   GCA_014885115.2   \n",
       "2  http://edamontology.org/operation_0362   GCA_037356235.1   \n",
       "3  http://edamontology.org/operation_0362   GCA_019693545.1   \n",
       "4  http://edamontology.org/operation_0362   GCA_036939335.1   \n",
       "\n",
       "                                     track_file_name  \\\n",
       "0  https://zenodo.org/records/14040072/files/Aste...   \n",
       "1  https://zenodo.org/records/14040072/files/Aste...   \n",
       "2  https://zenodo.org/records/14040072/files/Bact...   \n",
       "3  https://zenodo.org/records/14040072/files/Chae...   \n",
       "4  https://zenodo.org/records/14040072/files/Cont...   \n",
       "\n",
       "                                      track_file_url  \\\n",
       "0  https://zenodo.org/records/14040072/files/Aste...   \n",
       "1  https://zenodo.org/records/14040072/files/Aste...   \n",
       "2  https://zenodo.org/records/14040072/files/Bact...   \n",
       "3  https://zenodo.org/records/14040072/files/Chae...   \n",
       "4  https://zenodo.org/records/14040072/files/Cont...   \n",
       "\n",
       "                     track_file_format  \\\n",
       "0  http://edamontology.org/format_1975   \n",
       "1  http://edamontology.org/format_1975   \n",
       "2  http://edamontology.org/format_1975   \n",
       "3  http://edamontology.org/format_1975   \n",
       "4  http://edamontology.org/format_1975   \n",
       "\n",
       "                               track_feature  \\\n",
       "0  http://purl.obolibrary.org/obo/SO_0000704   \n",
       "1  http://purl.obolibrary.org/obo/SO_0000704   \n",
       "2  http://purl.obolibrary.org/obo/SO_0000704   \n",
       "3  http://purl.obolibrary.org/obo/SO_0000704   \n",
       "4  http://purl.obolibrary.org/obo/SO_0000704   \n",
       "\n",
       "  experiment_annotation_method_name  \\\n",
       "0                         [BRAKER3]   \n",
       "1                         [BRAKER2]   \n",
       "2                         [BRAKER3]   \n",
       "3                         [BRAKER3]   \n",
       "4                         [BRAKER3]   \n",
       "\n",
       "                       experiment_evidence_global_id  \\\n",
       "0  [SRR5749612, None, Viridiplantae+Stramenopiles...   \n",
       "1  [None, None, Viridiplantae+Stramenopiles (http...   \n",
       "2  [SRR18733610, None, Viridiplantae+Stramenopile...   \n",
       "3  [SRR8647946,SRR8647948,SRR8647949,SRR8647950,S...   \n",
       "4  [SRR18733581, None, Viridiplantae+Stramenopile...   \n",
       "\n",
       "             experiment_compute_protocol_description  \\\n",
       "0  https://docs.google.com/spreadsheets/d/1_1wKib...   \n",
       "1  https://docs.google.com/spreadsheets/d/1_1wKib...   \n",
       "2  https://docs.google.com/spreadsheets/d/1_1wKib...   \n",
       "3  https://docs.google.com/spreadsheets/d/1_1wKib...   \n",
       "4  https://docs.google.com/spreadsheets/d/1_1wKib...   \n",
       "\n",
       "                       track_evaluation_method_value  \\\n",
       "0  [C:95.0%[S:94.0%,D:1.0%],F:1.0%,M:4.0%,n:100,E...   \n",
       "1  [C:92.0%[S:92.0%,D:0.0%],F:1.0%,M:7.0%,n:100,E...   \n",
       "2  [C:95.0%[S:95.0%,D:0.0%],F:2.0%,M:3.0%,n:100,E...   \n",
       "3  [C:97.0%[S:96.0%,D:1.0%],F:1.0%,M:2.0%,n:100, ...   \n",
       "4  [C:99.0%[S:98.0%,D:1.0%],F:0.0%,M:1.0%,n:100,E...   \n",
       "\n",
       "   experiment_evidence_evidence_type  \\\n",
       "0  [RNA, proteins, protein database]   \n",
       "1  [RNA, proteins, protein database]   \n",
       "2  [RNA, proteins, protein database]   \n",
       "3  [RNA, proteins, protein database]   \n",
       "4  [RNA, proteins, protein database]   \n",
       "\n",
       "                        track_evaluation_method_name  \n",
       "0  [BUSCO scores genome level, BUSCO scores prote...  \n",
       "1  [BUSCO scores genome level, BUSCO scores prote...  \n",
       "2  [BUSCO scores genome level, BUSCO scores prote...  \n",
       "3  [BUSCO scores genome level, BUSCO scores prote...  \n",
       "4  [BUSCO scores genome level, BUSCO scores prote...  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
